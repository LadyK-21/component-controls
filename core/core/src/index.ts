export * from './build';
export * from './components';
export * from './configuration';
export * from './controls';
export * from './controls-utils';
export * from './deepMerge';
export * from './propsInfo';
export * from './document';
export * from './document-utils';
export * from './utility';
export { randomizeData, canRandomizeControl } from './controls-randomize';
export * from './controls-smart';
export * from './search';
export * from './source';
export * from './faker';

export type SetControlValueFn = (
  storyId: string,
  propertyName: string | undefined,
  value: any,
) => void;
export type ClickControlFn = (storyId: string, propertyName: string) => void;

export const normalizePath = (fillePath: string): string =>
  process.platform === 'win32' ? fillePath.replace(/\\/g, '\\\\') : fillePath;
