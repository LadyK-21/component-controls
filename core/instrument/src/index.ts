import fs from 'fs';
import mdx from '@mdx-js/mdx';
import matter from 'gray-matter';
import { File } from '@babel/types';
import traverse from '@babel/traverse';
import generate from '@babel/generator';
import {
  Story,
  Document,
  getASTSource,
  deepmerge,
} from '@component-controls/core';

import { extractCSFStories } from './babel/esm-stories';
import { extractMDXStories } from './babel/mdx-stories';
import { transformMDXAttributes } from './babel/transform-ast-tree';
import { extractStoreComponent } from './babel/extract-component';
import { packageInfo } from './misc/package-info';
import { extractStoryExports } from './misc/mdx-exports';
import { prettify } from './misc/prettify';
import { parseFile } from './misc/ast_store';
import { readSourceFile } from './misc/source-options';
import { getFileDates } from './misc/file-info';
import {
  LoadingDocStore,
  InstrumentOptions,
  ParserOptions,
  defaultParserOptions,
  ResolveOptions,
  defaultResolveOptions,
  ComponentOptions,
  defaultComponentOptions,
  StoriesOptions,
  defaultStoriesOptions,
  MDXOptions,
  defaultMDXOptions,
  ParseStorieReturnType,
} from './types';

export * from './types';
export { getComponentProps } from './misc/props-info';
export { getFileIinfo } from './misc/file-info';
export { prettify };

type TraverseFn = (
  ast: File,
  options: Required<InstrumentOptions>,
  more: { source: string; filePath: string },
) => ParseStorieReturnType | undefined;

/**
 * Instrument a source file, with options
 * @param code The source code
 * @param traverseFn A traverse function. can be different for MDX and ESM story files
 * @param originalSource If the source was modified (ie mdx compiler)
 * @param filePath file name with fiull path
 * @param options Instrument options
 */
const parseSource = async (
  code: string,
  traverseFn: TraverseFn,
  originalSource: string,
  filePath: string,
  options: Required<InstrumentOptions>,
): Promise<ParseStorieReturnType | undefined> => {
  const source = prettify(code, options.prettier, filePath);
  const { ast } = parseFile(filePath, options.parser, source);

  const store = traverseFn(ast, options, { source, filePath });
  if (!store) {
    return undefined;
  }
  if (store.doc) {
    const doc = store.doc;
    if (doc.draft === true) {
      if (process.env.NODE_ENV !== 'development') {
        return undefined;
      }
    }
    const saveSource = readSourceFile(
      options.stories.sourceFiles,
      originalSource,
      doc.title,
      filePath,
    );
    if (saveSource) {
      doc.source = saveSource;
    }
  }
  await extractStoreComponent(store, filePath, source, options);
  const doc: Document | undefined = store.doc;
  if (doc && store.stories) {
    const storyPackage = await packageInfo(
      doc.title,
      filePath,
      options.stories.package,
    );
    if (storyPackage) {
      store.packages[storyPackage.fileHash] = storyPackage;
      doc.package = storyPackage.fileHash;
    }
    const dates = await getFileDates(filePath);
    doc.dateModified = doc.dateModified || dates.dateModified;
    doc.date = doc.date || dates.dateCreated;
  }
  for (const key of Object.keys(store.stories)) {
    const story: Story = store.stories[key];
    if (!story.source) {
      story.source = getASTSource(source, story.loc);
    }
  }
  return store;
};

export { LoadingDocStore };
export type ParseStoriesReturnType = { transformed: string } & Partial<
  LoadingDocStore
>;
/**
 * Parse and instrument a javascript, typescript or MDX file of stories
 * @param source Source of the file to be instrumented
 * @param filePath Resolved file path name.
 * @param options Instrumenting options
 */

export const parseStories = async (
  filePath: string,
  fileSource?: string,
  options?: InstrumentOptions,
): Promise<ParseStoriesReturnType> => {
  const source = fileSource || fs.readFileSync(filePath, 'utf8');
  const {
    parser: parserOptions = {},
    prettier: prettierOptions = {},
    resolver: resolveOptions = {},
    components: componentOptions = {},
    stories: storiesOptions = {},
    mdx: mdxOptions = {},
    propsLoaders,
  } = options || {};

  const mergedOptions: Required<InstrumentOptions> = {
    parser: deepmerge<ParserOptions>(defaultParserOptions, parserOptions),
    resolver: deepmerge<ResolveOptions>(defaultResolveOptions, resolveOptions),
    prettier: prettierOptions,
    components: deepmerge<ComponentOptions>(
      defaultComponentOptions,
      componentOptions,
    ),
    stories: deepmerge<StoriesOptions>(defaultStoriesOptions, storiesOptions),
    mdx: deepmerge<MDXOptions>(defaultMDXOptions, mdxOptions),
    propsLoaders: propsLoaders || [],
  };
  let code: string;
  const {
    test,
    renderer,
    transformMDX,
    storybookExports = false,
    ...otherMDXOptions
  } = mergedOptions.mdx;
  if (test && filePath.match(test)) {
    const { data, content } = matter(source);
    code = await mdx(content, {
      filepath: filePath,
      ...otherMDXOptions,
    });
    const { ast } = parseFile(filePath, mergedOptions.parser, code, false);

    if (transformMDX) {
      //second pass transform - inject any necessary attributes
      traverse(ast, transformMDXAttributes(code));
      ({ code } = generate(ast, {
        retainFunctionParens: true,
        retainLines: false,
      }));
    }
    const store = await parseSource(
      code,
      extractMDXStories(data),
      content,
      filePath,
      mergedOptions,
    );
    if (!store) {
      return {
        transformed: code,
      };
    }
    let transformed;
    if (transformMDX) {
      const { exports } = store || {};
      const exportsSource = extractStoryExports(storybookExports, exports);
      transformed = `${renderer}\n${code}\n${exportsSource}`;
    } else {
      transformed = source;
    }
    return {
      transformed,
      ...store,
    };
  } else {
    const store = await parseSource(
      source,
      extractCSFStories,
      source,
      filePath,
      mergedOptions,
    );
    return {
      transformed: source,
      ...store,
    };
  }
};
