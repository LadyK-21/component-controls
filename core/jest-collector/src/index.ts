import { runCLI } from 'jest';
import { Config } from '@jest/types';
import { AssertionResult } from '@jest/test-result';
import { FileCoverage, CoverageSummary } from 'istanbul-lib-coverage';

export interface JestResults {
  /**
   * test results
   */
  testResults: AssertionResult[];
  /**
   * coverage summary data, by file
   */
  coverage: Record<string, CoverageSummary>;
}
export const run = async (
  testFile: string,
  testFolder: string,
): Promise<JestResults> => {
  const runResults = await runCLI(
    ({
      $0: testFile,
      roots: ['.'],
      silent: true,
      verbose: false,
      reporters: [],
      collectCoverage: true,
      coverageReporters: ['none'],
      watchman: false,
    } as unknown) as Config.Argv,
    [testFolder],
  );
  const cov = runResults.results.coverageMap;
  const result: JestResults = {
    testResults: runResults.results.testResults[0].testResults,
    coverage: {},
  };
  if (cov) {
    Object.keys(cov.data).forEach(file => {
      const fc = cov.data[file] as FileCoverage;
      result.coverage[file] = fc.toSummary();
    });
  }

  return result;
};
