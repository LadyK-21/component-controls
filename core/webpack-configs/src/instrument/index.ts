import { Configuration } from 'webpack';

export const instrument: Configuration = {
  module: {
    rules: [
      {
        test: /\.(ts|tsx)$/,
        use: [
          {
            loader: 'babel-loader',
            options: {
              presets: [['react-app', { flow: false, typescript: true }]],
            },
          },
        ],
      },
      {
        test: /\.mdx$/i,
        exclude: [/node_modules/],
        loader: 'babel-loader',
        options: {
          presets: [
            ['@babel/preset-env', { modules: 'commonjs' }],
            '@babel/preset-react',
          ],
        },
      },
      {
        test: /\.mdx$/i,
        exclude: [/node_modules/],
        loader: '@component-controls/loader/loader',
        enforce: 'pre',
        options: {
          mdx: {
            transformMDX: true,
            storybookExports: true,
          },
        },
      },
      {
        test: /\.(story|stories).(js|jsx|ts|tsx)$/,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        enforce: 'pre',
      },
    ],
  },
  resolve: {
    extensions: ['.ts', '.tsx', '.js', '.jsx'],
  },
};
