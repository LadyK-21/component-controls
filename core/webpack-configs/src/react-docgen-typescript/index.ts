import { Configuration } from 'webpack';

export const reactDocgenTypescript: Configuration = {
  module: {
    rules: [
      {
        test: /\.mdx$/i,
        exclude: [/node_modules/],
        loader: '@component-controls/loader/loader',
        enforce: 'pre',
        options: {
          propsLoaders: [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx)$/,
            },
            {
              name: '@component-controls/react-docgen-typescript-info',
              test: /\.(ts|tsx)$/,
            },
          ],
        },
      },

      {
        test: /\.(story|stories).(js|jsx|ts|tsx|mdx)$/i,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        enforce: 'pre',
        options: {
          propsLoaders: [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx)$/,
            },
            {
              name: '@component-controls/react-docgen-typescript-info',
              test: /\.(ts|tsx)$/,
            },
          ],
        },
      },
    ],
  },
};
