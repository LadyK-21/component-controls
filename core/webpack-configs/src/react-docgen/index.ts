import { Configuration } from 'webpack';

export const reactDocgen: Configuration = {
  module: {
    rules: [
      {
        test: /\.mdx$/i,
        exclude: [/node_modules/],
        loader: '@component-controls/loader/loader',
        enforce: 'pre',
        options: {
          propsLoaders: [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx|ts|tsx)$/,
            },
          ],
        },
      },
      {
        test: /\.(story|stories).(js|jsx|ts|tsx)$/,
        loader: '@component-controls/loader/loader',
        exclude: [/node_modules/],
        enforce: 'pre',
        options: {
          propsLoaders: [
            {
              name: '@component-controls/react-docgen-info',
              test: /\.(js|jsx|ts|tsx)$/,
            },
          ],
        },
      },
    ],
  },
};
