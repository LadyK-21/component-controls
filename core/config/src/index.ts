import * as path from 'path';
import * as fs from 'fs';

import { sync as globSync } from 'glob';
import globBase from 'glob-base';
import { makeRe } from 'micromatch';

import yargs from 'yargs';
import { BuildConfiguration } from '@component-controls/core';

export const buildConfigFileNames = [
  'buildtime.js',
  'build.js',
  'main.js',
  'doczrc.js',
];

export const optionsFileNames = [
  'runtime.js',
  'options.js',
  'runtime.ts',
  'runtime.tsx',
  'options.ts',
  'options.tsx',
];
export interface ConfigurationResult {
  config: BuildConfiguration;
  configPath: string;
  optionsFilePath?: string;
}
/**
 * return the configuration folder from command-line parameters
 * command line accepts -c/ -config parameter for config path
 * the config file is assumed named main.js/main.ts
 * @param baseFolder project folder to start the search with
 */
export const getConfigurationArg = (
  args: string[] = process.argv,
): string | undefined => {
  const argv = yargs(args)
    .option('config', {
      alias: 'c',
      description: 'configuration folder',
      type: 'string',
    })
    .help()
    .alias('help', 'h').argv;
  return argv.config;
};

/**
 *  given a base project folder and a configuration folder, returns the configuration file
 *
 * @param baseFolder project folder to start the search with
 * @param configFolder folder where the configuration file is located
 * @param args optional arguments
 */
export const loadConfiguration = (
  baseFolder: string,
  configFolder?: string,
  args?: string[],
  defaultConfigPath?: string,
): ConfigurationResult | undefined => {
  const folder = configFolder ?? getConfigurationArg(args) ?? defaultConfigPath;
  const configPath = folder ? path.resolve(baseFolder, folder) : baseFolder;
  const hasConfigFolder = fs.existsSync(configPath);
  if (!hasConfigFolder) {
    console.warn('configuration folder not found', configPath);
  }
  const allFiles = hasConfigFolder ? fs.readdirSync(configPath) : [];
  const buildConfigFile = allFiles.find(file =>
    buildConfigFileNames.includes(file.toLowerCase()),
  );
  const optionsFile = allFiles.find(file =>
    optionsFileNames.includes(file.toLowerCase()),
  );
  if (buildConfigFile) {
    const buildPath = path.resolve(configPath, buildConfigFile);
    let config = require('esm')(module)(buildPath);
    if (
      !config ||
      (typeof config === 'object' && Object.keys(config).length === 0)
    ) {
      config = require(buildPath);
    }
    return {
      config: config.default || config,
      optionsFilePath: optionsFile
        ? path.resolve(configPath, optionsFile)
        : undefined,
      configPath,
    };
  }
  return undefined;
};

//fix for sb5 issue handling glob
const fixGlob = (golbExpr: string): string => golbExpr.replace('.(', '.@(');

/**
 * find the story files out of a configuration file
 * using glob for the regex file search
 */
export const extractDocuments = ({
  config,
  configPath,
}: ConfigurationResult): string[] | undefined => {
  const stories = config ? config.stories || config.files : undefined;
  const files = typeof stories === 'string' ? [stories] : stories;
  const documents = files
    ? files.reduce((acc: string[], storyRg: string) => {
        const matches = globSync(path.resolve(configPath, fixGlob(storyRg)));
        if (!matches.length) {
          throw new Error(`${storyRg} did not match any files`);
        }
        const ignore = config.ignore || [];
        const files = matches.filter(filename => {
          const basename = path.basename(filename).toLowerCase();
          return !ignore.includes(basename);
        });
        return [...acc, ...files];
      }, [])
    : undefined;
  return documents;
};

/**
 * from the glob list of documents, extract require.context array of props
 */
export interface RequireContextProps {
  directory: string;
  useSubdirectories: boolean;
  regExp: RegExp;
}
/**
 * convert glob patters from config file into require.context objects
 */
export const configRequireContext = ({
  config,
  configPath,
}: ConfigurationResult): RequireContextProps[] | undefined => {
  const stories = config ? config.stories || config.files : undefined;
  const files = typeof stories === 'string' ? [stories] : stories;

  const contexts = files
    ? files.reduce((acc: RequireContextProps[], storyRg: string) => {
        const base = globBase(fixGlob(storyRg));
        const useSubdirectories = base.glob.startsWith('**');
        const glob = useSubdirectories ? base.glob.substr(3) : base.glob;
        const regExp = base.isGlob
          ? RegExp(makeRe(glob).source.substr(1))
          : new RegExp(glob);
        return [
          ...acc,
          {
            directory: path.resolve(configPath, base.base),
            useSubdirectories,
            regExp,
          },
        ];
      }, [])
    : undefined;
  return contexts;
};

/**
 * merge a configuration passed through cli or tools, with the build configration from the config path
 */
export const mergeBuildConfiguration = (
  config: BuildConfiguration,
): BuildConfiguration => {
  const { configPath } = config;
  const buildConfig = loadConfiguration(process.cwd(), configPath);
  return { ...buildConfig?.config, ...config };
};
