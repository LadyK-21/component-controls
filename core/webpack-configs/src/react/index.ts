import path from 'path';
import { RuleSetLoader } from 'webpack';
import ExtractCssPlugin from 'extract-css-chunks-webpack-plugin';
import OptimizeCssAssetsWebpackPlugin from 'optimize-css-assets-webpack-plugin';

import {
  PresetType,
  BuildProps,
  defCssFileName,
  customLoaderOptions,
} from '@component-controls/core';
import { findUpFile } from '@component-controls/core/node-utils';

export const react: PresetType = (options: BuildProps) => {
  const isProd = process.env.NODE_ENV === 'production';
  const cssLoaders: RuleSetLoader[] = [];
  const postcssOptions = customLoaderOptions(options, 'postcss-loader', {});
  const postCssOptionsFile = findUpFile(process.cwd(), 'postcss.config.js');
  const hasPostCss = Object.keys(postcssOptions).length || postCssOptionsFile;
  if (hasPostCss && !((postcssOptions as any).disable === true)) {
    cssLoaders.push({
      loader: 'postcss-loader',
      options: {
        postcssOptions:
          typeof postCssOptionsFile === 'string'
            ? require(postCssOptionsFile)
            : undefined,
        sourceMap: true,
      },
    });
  }
  const result: PresetType = {
    plugins: [
      new ExtractCssPlugin({
        filename: options.cssFileName || defCssFileName,
      }),
    ],
    optimization: { minimizer: [] },
    performance: { hints: false },
    module: {
      rules: [
        {
          test: /\.(js|jsx)$/,
          exclude: [/node_modules/],
          use: [
            {
              loader: 'babel-loader',
              options: {
                presets: [
                  [
                    '@babel/preset-env',
                    {
                      targets: {
                        browsers: ['last 5 versions', 'ie >= 9'],
                        node: 'current',
                      },
                      modules: 'commonjs',
                      useBuiltIns: 'usage',
                      corejs: 3,
                    },
                  ],
                  ['@babel/preset-react', { runtime: 'automatic' }],
                ],
              },
            },
          ],
        },
        {
          test: /\.(ts|tsx)$/,
          exclude: [/node_modules/],
          use: [
            {
              loader: 'babel-loader',
              options: {
                presets: [['react-app', { flow: false, typescript: true }]],
              },
            },
          ],
        },
        {
          test: /\.(md|mdx)$/i,
          exclude: [/node_modules/],
          loader: 'babel-loader',
          options: {
            presets: [
              [
                '@babel/preset-env',
                {
                  targets: {
                    browsers: ['last 5 versions', 'ie >= 9'],
                    node: 'current',
                  },
                  modules: 'commonjs',
                  useBuiltIns: 'usage',
                  corejs: 3,
                },
              ],
              '@babel/preset-react',
            ],
          },
        },
        {
          test: /\.(eot|md|svg|ico|jpg|jpeg|png|gif|ttf|woff|woff2|pdf|mp4|web|wav|mp3|m4a|aac|oga)$/i,
          exclude: [/node_modules/],
          loader: 'url-loader',
          options: customLoaderOptions(options, 'url-loader', {
            limit: 25000,
            name: '[name].[ext]',
            publicPath: '/static',
            outputPath: path.relative(
              options?.distFolder || process.cwd(),
              path.resolve(options?.staticFolder || process.cwd()),
            ),
          }),
        },
        {
          test: /\.(woff(2)?|ttf|eot|svg)(\?v=\d+\.\d+\.\d+)?$/,
          use: [
            {
              loader: 'file-loader',
              options: customLoaderOptions(options, 'file-loader', {
                name: '[name].[ext]',
                outputPath: path.relative(
                  options?.distFolder || process.cwd(),
                  path.resolve(options?.staticFolder || process.cwd()),
                ),
              }),
            },
          ],
        },
        {
          test: /\.(css|sass|scss|less)$/i,
          use: [
            // Creates `style` nodes from JS strings
            // will export to a consolidated css file
            ExtractCssPlugin.loader,
            {
              // Translates CSS into CommonJS
              loader: 'css-loader',
              options: customLoaderOptions(options, 'css-loader', {
                sourceMap: true,
              }),
            },
            ...cssLoaders,
            {
              // Compiles Sass to CSS
              loader: 'sass-loader',
              options: customLoaderOptions(options, 'sass-loader', {
                sourceMap: true,
              }),
            },
            {
              // Compiles less to CSS
              loader: 'less-loader',
              options: customLoaderOptions(options, 'less-loader', {
                sourceMap: true,
              }),
            },
          ],
        },
        {
          test: /\.txt$/i,
          use: [
            {
              loader: 'raw-loader',
              options: customLoaderOptions(options, 'raw-loader', {}),
            },
          ],
        },

        {
          test: /\.(md|mdx)$/i,
          exclude: [/node_modules/],
          loader: '@component-controls/loader/loader',
          enforce: 'pre',
          options: {
            mdx: {
              transformMDX: true,
            },
          },
        },
        {
          test: /\.(story|stories|doc|docs).(js|jsx|ts|tsx)$/,
          loader: '@component-controls/loader/loader',
          exclude: [/node_modules/],
          enforce: 'pre',
        },
      ],
    },
    resolve: {
      extensions: ['.ts', '.tsx', '.js', '.jsx'],
    },
  };
  if (isProd) {
    result.optimization = {
      minimizer: [
        new OptimizeCssAssetsWebpackPlugin({
          cssProcessorOptions: {
            discardComments: { removeAll: true },
          },
        }),
      ],
    };
  }
  return result;
};
