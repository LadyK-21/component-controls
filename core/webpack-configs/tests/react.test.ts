import { mergeWebpackConfig } from '../src';
describe('merge instrument and react-docgen-typescript', () => {
  it('get config', () => {
    const config = mergeWebpackConfig(
      undefined,
      ['react', 'react-docgen-typescript'],
      {},
    );
    expect(config).toMatchObject({
      plugins: [
        {
          _sortedModulesCache: {},
          options: {
            filename: 'component-controls.css',
            ignoreOrder: false,
            experimentalUseImportModule: false,
            chunkFilename: '[id].component-controls.css',
          },
          runtimeOptions: {
            insert: undefined,
            linkType: 'text/css',
            attributes: undefined,
          },
        },
      ],
      optimization: {
        minimizer: [],
      },
      performance: {
        hints: false,
      },
      module: {
        rules: [
          {
            test: {},
            exclude: [{}],
            use: [
              {
                loader: 'babel-loader',
                options: {
                  presets: [
                    [
                      '@babel/preset-env',
                      {
                        targets: {
                          browsers: ['last 5 versions', 'ie >= 9'],
                          node: 'current',
                        },
                        modules: 'commonjs',
                        useBuiltIns: 'usage',
                        corejs: 3,
                      },
                    ],
                    [
                      '@babel/preset-react',
                      {
                        runtime: 'classic',
                      },
                    ],
                  ],
                },
              },
            ],
          },
          {
            test: {},
            exclude: [{}],
            use: [
              {
                loader: 'babel-loader',
                options: {
                  presets: [
                    [
                      'react-app',
                      {
                        flow: false,
                        typescript: true,
                      },
                    ],
                  ],
                },
              },
            ],
          },
          {
            test: {},
            exclude: [{}],
            loader: 'babel-loader',
            options: {
              presets: [
                [
                  '@babel/preset-env',
                  {
                    targets: {
                      browsers: ['last 5 versions', 'ie >= 9'],
                      node: 'current',
                    },
                    modules: 'commonjs',
                    bugfixes: true,
                    useBuiltIns: 'usage',
                    corejs: 3,
                  },
                ],
                '@babel/preset-react',
              ],
            },
          },
          {
            test: {},
            exclude: [{}],
            loader: 'url-loader',
            options: {
              limit: 24576,
              name: '[name].[ext]',
              publicPath: '/static',
              outputPath: '',
            },
          },
          {
            test: {},
            use: [
              {
                loader: 'file-loader',
                options: {
                  name: '[name].[ext]',
                  outputPath: '',
                },
              },
            ],
          },
          {
            test: {},
            use: [
              {
                options: {},
              },
              {
                loader: 'css-loader',
                options: {
                  sourceMap: true,
                },
              },
              {
                loader: 'sass-loader',
                options: {
                  sourceMap: true,
                },
              },
              {
                loader: 'less-loader',
                options: {
                  sourceMap: true,
                },
              },
            ],
          },
          {
            test: {},
            use: [
              {
                loader: 'raw-loader',
                options: {},
              },
            ],
          },
          {
            test: {},
            use: [
              {
                loader: 'raw-loader',
                options: {},
              },
            ],
          },
          {
            test: {},
            exclude: [{}, {}],
            loader: '@component-controls/loader/loader',
            enforce: 'pre',
            options: {
              propsLoaders: [
                {
                  name: '@component-controls/react-docgen-info',
                  test: {},
                },
                {
                  name: '@component-controls/react-docgen-typescript-info',
                  test: {},
                },
              ],
              mdx: {
                transformMDX: true,
              },
            },
          },
          {
            test: {},
            loader: '@component-controls/loader/loader',
            exclude: [{}, {}],
            enforce: 'pre',
            options: {
              propsLoaders: [
                {
                  name: '@component-controls/react-docgen-info',
                  test: {},
                },
                {
                  name: '@component-controls/react-docgen-typescript-info',
                  test: {},
                },
              ],
            },
          },
        ],
      },
      resolve: {
        extensions: ['.ts', '.tsx', '.js', '.jsx'],
      },
    });
  });
  it('custom loader options', () => {
    const config = mergeWebpackConfig(
      undefined,
      ['react', 'react-docgen-typescript'],
      {
        loaders: {
          'css-loader': {
            modules: {
              localIdentName: '[name].[local].[hash]',
            },
          },
        },
      },
    );
    let cssLoader;
    for (const rule of config.module.rules) {
      if (Array.isArray(rule.use)) {
        const use = rule.use.find(
          use => typeof use === 'object' && use.loader === 'css-loader',
        );
        if (use) {
          cssLoader = use;
          break;
        }
      }
    }
    expect(cssLoader.options).toMatchObject({
      modules: { localIdentName: '[name].[local].[hash]' },
    });
  });
});
