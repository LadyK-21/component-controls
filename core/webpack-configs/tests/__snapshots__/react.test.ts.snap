// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`merge instrument and react-docgen-typescript get config 1`] = `
Object {
  "module": Object {
    "rules": Array [
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                  Object {
                    "corejs": 3,
                    "modules": "commonjs",
                    "targets": Object {
                      "browsers": Array [
                        "last 5 versions",
                        "ie >= 9",
                      ],
                      "node": "current",
                    },
                    "useBuiltIns": "usage",
                  },
                ],
                Array [
                  "@babel/preset-react",
                  Object {
                    "runtime": "automatic",
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
        "use": Array [
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "react-app",
                  Object {
                    "flow": false,
                    "typescript": true,
                  },
                ],
              ],
            },
          },
        ],
      },
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "loader": "babel-loader",
        "options": Object {
          "presets": Array [
            Array [
              "@babel/preset-env",
              Object {
                "corejs": 3,
                "modules": "commonjs",
                "targets": Object {
                  "browsers": Array [
                    "last 5 versions",
                    "ie >= 9",
                  ],
                  "node": "current",
                },
                "useBuiltIns": "usage",
              },
            ],
            "@babel/preset-react",
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/i,
      },
      Object {
        "exclude": Array [
          /node_modules/,
        ],
        "loader": "url-loader",
        "options": Object {
          "limit": 25000,
          "name": "[name].[ext]",
          "outputPath": "",
          "publicPath": "/static",
        },
        "test": /\\\\\\.\\(eot\\|md\\|svg\\|ico\\|jpg\\|jpeg\\|png\\|gif\\|ttf\\|woff\\|woff2\\|pdf\\|mp4\\|web\\|wav\\|mp3\\|m4a\\|aac\\|oga\\)\\$/i,
      },
      Object {
        "test": /\\\\\\.\\(woff\\(2\\)\\?\\|ttf\\|eot\\|svg\\)\\(\\\\\\?v=\\\\d\\+\\\\\\.\\\\d\\+\\\\\\.\\\\d\\+\\)\\?\\$/,
        "use": Array [
          Object {
            "loader": "file-loader",
            "options": Object {
              "name": "[name].[ext]",
              "outputPath": "",
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.\\(css\\|sass\\|scss\\|less\\)\\$/i,
        "use": Array [
          "/Users/atanasster/component-controls/node_modules/extract-css-chunks-webpack-plugin/dist/loader.js",
          Object {
            "loader": "css-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "sass-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
          Object {
            "loader": "less-loader",
            "options": Object {
              "sourceMap": true,
            },
          },
        ],
      },
      Object {
        "test": /\\\\\\.txt\\$/i,
        "use": Array [
          Object {
            "loader": "raw-loader",
            "options": Object {},
          },
        ],
      },
      Object {
        "enforce": "pre",
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "mdx": Object {
            "transformMDX": true,
          },
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(md\\|mdx\\)\\$/i,
      },
      Object {
        "enforce": "pre",
        "exclude": Array [
          /node_modules/,
          /node_modules/,
        ],
        "loader": "@component-controls/loader/loader",
        "options": Object {
          "propsLoaders": Array [
            Object {
              "name": "@component-controls/react-docgen-info",
              "test": /\\\\\\.\\(js\\|jsx\\)\\$/,
            },
            Object {
              "name": "@component-controls/react-docgen-typescript-info",
              "test": /\\\\\\.\\(ts\\|tsx\\)\\$/,
            },
          ],
        },
        "test": /\\\\\\.\\(story\\|stories\\|doc\\|docs\\)\\.\\(js\\|jsx\\|ts\\|tsx\\)\\$/,
      },
    ],
  },
  "optimization": Object {
    "minimizer": Array [],
  },
  "performance": Object {
    "hints": false,
  },
  "plugins": Array [
    ExtractCssChunksPlugin {
      "options": Object {
        "chunkFilename": "[id].component-controls.css",
        "filename": "component-controls.css",
        "ignoreOrder": false,
        "insert": null,
        "moduleFilename": [Function],
      },
    },
  ],
  "resolve": Object {
    "extensions": Array [
      ".ts",
      ".tsx",
      ".js",
      ".jsx",
    ],
  },
}
`;
